wgpu-hal/src/lib.rs

```diff
@@ -652,6 +656,10 @@ pub trait CommandEncoder: WasmNotSendSync + fmt::Debug {
     ///
     /// This puts this `CommandEncoder` in the "closed" state.
     ///
+    /// Implementations of this function must be idempotent, i.e.
+    /// if the function has just been called, calling it again should
+    /// not do anything.
```

Comment:
Not pushing null handles onto `CommandEncoder::discarded` seems clearly necessary, lest `begin_encoding` pop them back off and try to use them. And if it turns out that the easiest fix for some bug happens to make a `wgpu_hal` trait method idempotent, great.

But adding idempotency as a requirement for implementations of the `CommandEncoder::discard_encoding` trait method is not the right move. See the crate-level docs in `wgpu_hal/src/lib.rs` for a better sense of why. (Those docs only landed a few days ago, but they're just an expansion of the terse/cryptic docs that were there before.)

Reply:
Given #5566, please flip this around to say, instead, that callers *must not* assume that this function is idempotent.

The fact that the PR does make the Vulkan implementation idempotent doesn't change the caller's obligations.